text/0000-http-index.md

```diff
@@ -0,0 +1,124 @@
+
+## Bandwidth reduction
+
+Cargo supports HTTP/2, which handles many similar requests efficiently.
+
+All fetched dependency files can be cached, and refreshed using conditional HTTP requests (with `Etag` or `If-Modified-Since` headers), to avoid redownloading of files that haven't changed.
+
+Dependency files compress well. Currently the largest file of `rustc-ap-rustc_data_structures` compresses from 1MiB to 26KiB with Brotli. Many servers support transparently serving pre-compressed files (i.e. request for `/rustc-ap-rustc_data_structures` can be served from `rustc-ap-rustc_data_structures.gz` with an appropriate content encoding header), so the index can use high compression levels without increasing CPU cost of serving the files.
+
+### Optionally, a rotated incremental changelog
+
+To further reduce number requests needed to update the index, the index may maintain an append-only log of changes. For each change (crate version published or yanked), the log would append a line with: epoch number (explained below), last-modified timestamp, and the name of the changed crate, e.g.
+
+```
+1 2019-10-18 23:51:23 oxigen
+1 2019-10-18 23:51:25 linda
+1 2019-10-18 23:51:29 rv
+1 2019-10-18 23:52:00 anyhow
+1 2019-10-18 23:53:03 build_id
+1 2019-10-18 23:56:16 canonical-form
+1 2019-10-18 23:59:01 cotton
+1 2019-10-19 00:01:44 kg-utils
+1 2019-10-19 00:08:45 serde_traitobject
+```
+
+Because the log is append-only, the client can incrementally update it using a `Range` HTTP request. The client doesn't have to download the full log in order to start using it; it can download only an arbitrary fraction of it, up to the end of the file, which is straightforward with a `Range` request. When a crate is found in the log (searching from the end), and modification date is the same as modification date of crate's cached locally, the client won't have to make an HTTP request for the file.
+
+When the log grows too big, the epoch number can be incremented, and the log reset back to empty. The epoch number allows clients to detect that the log has been reset, even if the `Range` they requested happened to be valid for the new log file.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* A basic solution, without the incremental changelog, needs more requests and has higher latency to update the index. With the help of the incremental changelog, this is largely mitigated.
+* Performant implementation of this solution depends on making many small requests in parallel. This in practice requires HTTP/2 support on the server.
+* It's uncertain if GitHub pages can handle this many files and the amount of traffic they generate, so the index may need to be hosted elsewhere.
+* Since alternative registries are stable, the git-based protocol is stable, and can't be removed.
+
+# Rationale and alternatives
```

Comment:
is some form of `rsync` or similar not an alternative?

that maybe "lazily" (following the above on-demand of particular crates) but also maybe "simply" (in that rsync-ish things are well understood?) synchronises the full registry on request with plenty of compression and the like

Reply:
`rsync` is very very unsuitable for this:

* it requires a custom server-side daemon, so you can't run it off a cheap static file CDN.
* it creates checksums for all files on both sides, which means scanning of 40K files on each side
* it doesn't make the index on the client side any smaller, unless you try fetching subset of files, but then you're back to the same incremental solution as HTTP-based index, just built on top of a transport with higher overhead.